clang-tidy user options appear after '--'

<!--
  This is the template for reporting ALE bugs. Make sure you try updating ALE
  to a more recent version before reporting a bug. Look through existing bug
  reports for similar issues before reporting a new one. Don't leave comments
  about new bugs in the comment section for old issues.

  Make sure to try disabling other plugins and trying to repeat your bug before
  reporting it in ALE. Some times problems can arise when two plugins are used
  together, but often your issues might be problems with other plugins.
-->

## Information

VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Apr 10 2018 21:31:58)                                                                                                                                                 
Included patches: 1-1453                                                                                                                                                                                           

Operating System: Ubuntu 18.04.2 LTS

## What went wrong

<!-- Describe what went wrong here. Be specific. -->
In a project with no compilation database available, I added `let g:ale_cpp_clangtidy_options='--extra-arg=--std=c++17' ` to my config, but didn't appear to take effect. Looking at `ALEInfo` shows the clang-tidy command line ends with

```-- -extra-arg=-std=c++17```

which (I believe) needs to be 

```-extra-arg=-std=c++17 --```

## Reproducing the bug

<!-- Write a list of steps below. -->

1. In a project with no compilation database, open a header that requires C++17. Note horrible spew of error messages (clang defaults to a much earlier standard).

n.b. -- should be enough to have a header with just this line: 
```cpp
static_assert( __cplusplus >= 201703L, __FILE__ " requires C++ 17" );
```
2. let g:ale_cpp_clangtidy_options='--extra-arg=--std=c++17. Note spew of error messages continues.
3. Note the incorrect order of options in ALEInfo

### :ALEInfo

```
 Current Filetype: cpp
Available Linters: ['ccls', 'clang', 'clangcheck', 'clangd', 'clangtidy', 'clazy', 'cppcheck', 'cpplint', 'cquery', 'flawfinder', 'gcc']
   Linter Aliases:
'gcc' -> ['g++']
  Enabled Linters: ['ccls', 'clang', 'clangcheck', 'clangd', 'clangtidy', 'clazy', 'cppcheck', 'cpplint', 'cquery', 'flawfinder', 'gcc']
 Suggested Fixers: 
  'clang-format' - Fix C/C++ files with clang-format.
  'remove_trailing_lines' - Remove all blank lines at the end of a file.
  'trim_whitespace' - Remove all trailing whitespace characters at the end of every line.
  'uncrustify' - Fix C, C++, C#, ObjectiveC, ObjectiveC++, D, Java, Pawn, and VALA files with uncrustify.
 Linter Variables:

let g:ale_cpp_ccls_executable = 'ccls'
let g:ale_cpp_ccls_init_options = {}
let g:ale_cpp_clang_executable = 'clang++'
let g:ale_cpp_clang_options = '--std=c++17'
let g:ale_cpp_clangcheck_executable = 'clang-check'
let g:ale_cpp_clangcheck_options = '--extra-arg=--std=c++17'
let g:ale_cpp_clangd_executable = 'clangd'
let g:ale_cpp_clangd_options = ''
let g:ale_cpp_clangtidy_checks = []
let g:ale_cpp_clangtidy_executable = 'clang-tidy'
let g:ale_cpp_clangtidy_options = '--extra-arg=--std=c++17'
let g:ale_cpp_clazy_checks = ['level1']
let g:ale_cpp_clazy_executable = 'clazy-standalone'
let g:ale_cpp_clazy_options = ''
let g:ale_cpp_cppcheck_executable = 'cppcheck'
let g:ale_cpp_cppcheck_options = '--enable=style'
let g:ale_cpp_cpplint_executable = 'cpplint'
let g:ale_cpp_cpplint_options = ''
let g:ale_cpp_cquery_cache_directory = '/home/badams/.cache/cquery'
let g:ale_cpp_cquery_executable = 'cquery'
let g:ale_cpp_flawfinder_executable = 'flawfinder'
let g:ale_cpp_flawfinder_minlevel = 1
let g:ale_cpp_flawfinder_options = ''
let g:ale_cpp_gcc_executable = 'gcc'
let g:ale_cpp_gcc_options = '--std=c++17'
 Global Variables:

let g:ale_cache_executable_check_failures = v:null
let g:ale_change_sign_column_color = 0
let g:ale_command_wrapper = ''
let g:ale_completion_delay = 100
let g:ale_completion_enabled = 1
let g:ale_completion_max_suggestions = 50
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '[%linter%] %s [%severity%]'
let g:ale_echo_msg_info_str = 'Info'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {}
let g:ale_history_enabled = 1
let g:ale_history_log_output = 1
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_filetype_changed = 1
let g:ale_lint_on_insert_leave = 0
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'always'
let g:ale_linter_aliases = {}
let g:ale_linters = {}
let g:ale_linters_explicit = 0
let g:ale_list_vertical = 0
let g:ale_list_window_size = 10
let g:ale_loclist_msg_format = '[%linter%] %s [%severity%]'
let g:ale_max_buffer_history_size = 20
let g:ale_max_signs = -1
let g:ale_maximum_file_size = v:null
let g:ale_open_list = 0
let g:ale_pattern_options = v:null
let g:ale_pattern_options_enabled = v:null
let g:ale_set_balloons = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = '⇒'
let g:ale_sign_info = '→'
let g:ale_sign_offset = 1000000
let g:ale_sign_style_error = '⇒'
let g:ale_sign_style_warning = '→'
let g:ale_sign_warning = '→'
let g:ale_statusline_format = v:null
let g:ale_type_map = {}
let g:ale_use_global_executables = v:null
let g:ale_virtualtext_cursor = 0
let g:ale_warn_about_trailing_blank_lines = 1
let g:ale_warn_about_trailing_whitespace = 1
  Command History:

(executable check - success) clang++
(finished - exit code 0) ['/bin/bash', '-c', '''clang++'' -S -x c++ -fsyntax-only -iquote ''/home/badams/dev/frisbee/generated'' --std=c++17 - < ''/tmp/vl4ngMu/1/testschema.hpp''']

<<<OUTPUT STARTS>>>
<stdin>:1:9: warning: #pragma once in main file [-Wpragma-once-outside-header]
#pragma once
        ^
1 warning generated.
<<<OUTPUT ENDS>>>

(executable check - success) clang-check
(finished - exit code 0) ['/bin/bash', '-c', '''clang-check'' -analyze ''/home/badams/dev/frisbee/generated/testschema.hpp'' -extra-arg -Xclang -extra-arg -analyzer-output=text --extra-arg=--std=c++17']

<<<OUTPUT STARTS>>>
Error while trying to load a compilation database:
Could not auto-detect compilation database for file "/home/badams/dev/frisbee/generated/testschema.hpp"
No compilation database found in /home/badams/dev/frisbee/generated or any parent directory
fixed-compilation-database: Error while opening fixed database: No such file or directory
json-compilation-database: Error while opening JSON database: No such file or directory
Running without flags.
<<<OUTPUT ENDS>>>

(executable check - success) clang-tidy
(finished - exit code 1) ['/bin/bash', '-c', '''clang-tidy'' ''/home/badams/dev/frisbee/generated/testschema.hpp'' -- --extra-arg=--std=c++17']

<<<OUTPUT STARTS>>>
error: too many errors emitted, stopping now [clang-diagnostic-error]
error: unsupported option '--extra-arg=--std=c++17' [clang-diagnostic-error]
/home/badams/dev/frisbee/generated/testschema.hpp:17:1: error: static_assert failed "/home/badams/dev/frisbee/generated/testschema.hpp requires C++ 17" [clang-diagnostic-error]
static_assert( __cplusplus >= 201703L, __FILE__ " requires C++ 17" );
^
/home/badams/dev/frisbee/generated/testschema.hpp:18:15: warning: nested namespace definition is a C++17 extension; define each namespace separately [clang-diagnostic-c++17-extensions]
namespace test::testschema {
              ^~~~~~~~~~~~
               { namespace testschema
/home/badams/dev/frisbee/generated/testschema.hpp:27:20: warning: constexpr if is a C++17 extension [clang-diagnostic-c++17-extensions]
                if constexpr ( std::is_same_v< std::remove_cv_t< T >, char > ) {
                   ^
.... clip ....
           ^
<<<OUTPUT ENDS>>>

(executable check - failure) clazy-standalone
(executable check - success) cppcheck
(finished - exit code 1) ['/bin/bash', '-c', '''cppcheck'' -q --language=c++ --enable=style ''/tmp/vl4ngMu/2/testschema.hpp''']

<<<OUTPUT STARTS>>>
cppcheck: error: could not find or open any of the paths given.
<<<OUTPUT ENDS>>>

(executable check - failure) cpplint
(executable check - failure) flawfinder
(executable check - success) gcc
(finished - exit code 0) ['/bin/bash', '-c', '''gcc'' -S -x c++ -fsyntax-only -iquote ''/home/badams/dev/frisbee/generated'' --std=c++17 - < ''/tmp/vl4ngMu/3/testschema.hpp''']

<<<OUTPUT STARTS>>>
<stdin>:1:9: warning: #pragma once in main file
<<<OUTPUT ENDS>>>
```

--- a/ale_linters/c/clangtidy.vim
+++ b/ale_linters/c/clangtidy.vim
@@ -11,9 +11,12 @@ call ale#Set('c_clangtidy_executable', 'clang-tidy')
 " http://clang.llvm.org/extra/clang-tidy/checks/list.html
 
 call ale#Set('c_clangtidy_checks', [])
-" Set this option to manually set some options for clang-tidy.
+" Set this option to manually set some options for clang-tidy to use as compile
+" flags.
 " This will disable compile_commands.json detection.
 call ale#Set('c_clangtidy_options', '')
+" Set this option to manually set options for clang-tidy directly.
+call ale#Set('c_clangtidy_extra_options', '')
 call ale#Set('c_build_dir', '')
 
 function! ale_linters#c#clangtidy#GetCommand(buffer) abort
@@ -25,8 +28,12 @@ function! ale_linters#c#clangtidy#GetCommand(buffer) abort
     \   ? ale#Var(a:buffer, 'c_clangtidy_options')
     \   : ''
 
+    " Get the options to pass directly to clang-tidy
+    let l:extra_options = ale#Var(a:buffer, 'c_clangtidy_extra_options')
+
     return '%e'
     \   . (!empty(l:checks) ? ' -checks=' . ale#Escape(l:checks) : '')
+    \   . (!empty(l:extra_options) ? ' ' . ale#Escape(l:extra_options) : '')
     \   . ' %s'
     \   . (!empty(l:build_dir) ? ' -p ' . ale#Escape(l:build_dir) : '')
     \   . (!empty(l:options) ? ' -- ' . l:options : '')

--- a/ale_linters/cpp/clangtidy.vim
+++ b/ale_linters/cpp/clangtidy.vim
@@ -5,9 +5,12 @@
 call ale#Set('cpp_clangtidy_executable', 'clang-tidy')
 " Set this option to check the checks clang-tidy will apply.
 call ale#Set('cpp_clangtidy_checks', [])
-" Set this option to manually set some options for clang-tidy.
+" Set this option to manually set some options for clang-tidy to use as compile
+" flags.
 " This will disable compile_commands.json detection.
 call ale#Set('cpp_clangtidy_options', '')
+" Set this option to manually set options for clang-tidy directly.
+call ale#Set('cpp_clangtidy_extra_options', '')
 call ale#Set('c_build_dir', '')
 
 function! ale_linters#cpp#clangtidy#GetCommand(buffer) abort
@@ -19,8 +22,12 @@ function! ale_linters#cpp#clangtidy#GetCommand(buffer) abort
     \   ? ale#Var(a:buffer, 'cpp_clangtidy_options')
     \   : ''
 
+    " Get the options to pass directly to clang-tidy
+    let l:extra_options = ale#Var(a:buffer, 'cpp_clangtidy_extra_options')
+
     return '%e'
     \   . (!empty(l:checks) ? ' -checks=' . ale#Escape(l:checks) : '')
+    \   . (!empty(l:extra_options) ? ' ' . ale#Escape(l:extra_options) : '')
     \   . ' %s'
     \   . (!empty(l:build_dir) ? ' -p ' . ale#Escape(l:build_dir) : '')
     \   . (!empty(l:options) ? ' -- ' . l:options : '')

--- a/doc/ale-c.txt
+++ b/doc/ale-c.txt
@@ -156,7 +156,7 @@ g:ale_c_clangtidy_options                           *g:ale_c_clangtidy_options*
   Type: |String|
   Default: `''`
 
-  This variable can be changed to modify flags given to clang-tidy.
+  This variable can be changed to modify compiler flags given to clang-tidy.
 
   - Setting this variable to a non-empty string,
   - and working in a buffer where no compilation database is found using
@@ -167,6 +167,14 @@ g:ale_c_clangtidy_options                           *g:ale_c_clangtidy_options*
   Only set this option if you want to control compiler flags
   entirely manually, and no `compile_commands.json` file is in one
   of the |g:ale_c_build_dir_names| directories of the project tree.
+
+
+g:ale_c_clangtidy_extra_options               *g:ale_c_clangtidy_extra_options*
+                                              *b:ale_c_clangtidy_extra_options*
+  Type: |String|
+  Default: `''`
+
+  This variable can be changed to modify flags given to clang-tidy.
 
 
 ===============================================================================

--- a/doc/ale-cpp.txt
+++ b/doc/ale-cpp.txt
@@ -125,7 +125,7 @@ g:ale_cpp_clangtidy_options                       *g:ale_cpp_clangtidy_options*
   Type: |String|
   Default: `''`
 
-  This variable can be changed to modify flags given to clang-tidy.
+  This variable can be changed to modify compiler flags given to clang-tidy.
 
   - Setting this variable to a non-empty string,
   - and working in a buffer where no compilation database is found using
@@ -136,6 +136,14 @@ g:ale_cpp_clangtidy_options                       *g:ale_cpp_clangtidy_options*
   Only set this option if you want to control compiler flags
   entirely manually, and no `compile_commands.json` file is in one
   of the |g:ale_c_build_dir_names| directories of the project tree.
+
+
+g:ale_cpp_clangtidy_extra_options           *g:ale_cpp_clangtidy_extra_options*
+                                            *b:ale_cpp_clangtidy_extra_options*
+  Type: |String|
+  Default: `''`
+
+  This variable can be changed to modify flags given to clang-tidy.
 
 
 ===============================================================================

--- a/test/command_callback/test_c_clang_tidy_command_callback.vader
+++ b/test/command_callback/test_c_clang_tidy_command_callback.vader
@@ -29,6 +29,12 @@ Execute(You should be able to manually set compiler flags for clang-tidy):
   AssertLinter 'clang-tidy',
   \ ale#Escape('clang-tidy') . ' -checks=' . ale#Escape('*') . ' %s -- -Wall'
 
+Execute(You should be able to manually set flags for clang-tidy):
+  let b:ale_c_clangtidy_extra_options = '-config='
+
+  AssertLinter 'clang-tidy',
+  \ ale#Escape('clang-tidy') . ' ' . ale#Escape('-config=') . ' %s'
+
 Execute(The build directory should be configurable):
   let b:ale_c_clangtidy_checks = ['*']
   let b:ale_c_build_dir = '/foo/bar'

--- a/test/command_callback/test_clang_tidy_command_callback.vader
+++ b/test/command_callback/test_clang_tidy_command_callback.vader
@@ -29,6 +29,12 @@ Execute(You should be able to manually set compiler flags for clang-tidy):
   AssertLinter 'clang-tidy',
   \ ale#Escape('clang-tidy') . ' -checks=' . ale#Escape('*') . ' %s -- -Wall'
 
+Execute(You should be able to manually set flags for clang-tidy):
+  let b:ale_cpp_clangtidy_extra_options = '-config='
+
+  AssertLinter 'clang-tidy',
+  \ ale#Escape('clang-tidy') . ' ' . ale#Escape('-config=') . ' %s'
+
 Execute(The build directory should be configurable):
   let b:ale_cpp_clangtidy_checks = ['*']
   let b:ale_c_build_dir = '/foo/bar'

